### PR TITLE
gobject-introspection: fix SEGV on Apple Silicon (ARM64) Macs

### PR DESCRIPTION
**Pull request by mattrose**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
I know that seven patches is a lot to add, but this is the way the merge request was formatted upstream, and it's all necessary.  I would normally wait, but they're not planning on doing another release until February or March, and this stops a lot of python GTK projects from working on Apple Silicon

### CODE REVIEW

**Review by carlocab — COMMENTED** [review]
Thanks, @mattrose.

Can you amend your commit message to read `gobject-introspection: fix SEGV on Apple Silicon ARM64) Macs` (or similar)? We put the names of formulae first in our commit messages. (See the style guide linked in the PR template.)

Also, I'm wondering if these patches are better rolled into a single one and hosted at Homebrew/formula-patches.

Finally:

This seems like a pretty substantial patch that changes an API. (It doesn't look like it should break the ABI, but I can't say for sure.) That makes me question its suitability for a patch. (Indeed, the upstream MR bumps the minor version due to the API change.)

On the other hand, segfaults are not good at all. If we want to apply this patch, we should probably restrict it to the versions/arches that need it.

CC @Homebrew/core for thoughts (esp. @tschoonj)

**Comment by tschoonj** [comment]
@mattrose what is the link to the upstream gnome Gitlab MR? Thanks

File: Formula/gobject-introspection.rb [hunk 1]
```diff
@@ -41,6 +41,38 @@ class GobjectIntrospection < Formula
     sha256 "740c9fba499b1491689b0b1216f9e693e5cb35c9a8565df4314341122ce12f81"
   end
 
+  # Series of seven commits from MR-301 to Fix SEGV on Apple Silicon Macs
+  # See: https://github.com/Homebrew/homebrew-core/issues/88801
+  #      https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/301
```
Comment (SMillerDev) [inline Formula/gobject-introspection.rb:46]: @tschoonj

**Comment by tschoonj** [comment]
I am ok with this PR.

Instead of using the individual commits, you can use the combined patch: https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/301.patch

**Comment by fxcoudert** [comment]
@tschoonj part of that patch (1/8) changes the version of gobject-introspection from 1.70.1 to 1.71.0? Do we really want that? I think we should avoid changing the version…

**Comment by tschoonj** [comment]
> @tschoonj part of that patch (1/8) changes the version of gobject-introspection from 1.70.1 to 1.71.0? Do we really want that? I think we should avoid changing the version…

I was not aware of that. Better to use the individual commits then I guess.

**Comment by mattrose** [comment]
Thanks for the feed back.  My response in an hour or so.

**Comment by mattrose** [comment]
> Can you amend your commit message to read `gobject-introspection: fix SEGV on Apple Silicon ARM64) Macs` (or similar)? We put the names of formulae first in our commit messages. (See the style guide linked in the PR template.)

This has been done.

> 
> Also, I'm wondering if these patches are better rolled into a single one and hosted at Homebrew/formula-patches.

That would be fantastic.  This was just the best I could do with what I had available.  If you could do that, or point to instructions on how to submit to formula-patches, that would be great.

> 
> Finally:
> 
> This seems like a pretty substantial patch that changes an API. (It doesn't look like it should break the ABI, but I can't say for sure.) That makes me question its suitability for a patch. (Indeed, the upstream MR bumps the minor version due to the API change.)
> 
> On the other hand, segfaults are not good at all. If we want to apply this patch, we should probably restrict it to the versions/arches that need it.

I think there's consensus that this probaby needs to be patched.  It's actually a consequence of how libffi changed their API between minor versions, and this patch just deals with that.  As for restricting it to the versions and arches that need it, one of the patches itself does that restriction, so I don't think we need to duplicate that in homebrew.

Thanks again for looking at this.  It's always a joy to work with you guys to get stuff into Homebrew.

**Comment by mattrose** [comment]
> @tschoonj part of that patch (1/8) changes the version of gobject-introspection from 1.70.1 to 1.71.0? Do we really want that? I think we should avoid changing the version…

Also, brew audit doesn't like that because the MR patch can change, and says to use commit urls instead 😄

**Comment by MikeMcQuaid** [comment]
> On the other hand, segfaults are not good at all. If we want to apply this patch, we should probably restrict it to the versions/arches that need it.

Agreed.



> As for restricting it to the versions and arches that need it, one of the patches itself does that restriction, so I don't think we need to duplicate that in homebrew.

I think it's generally nicer to not apply patches when they aren't needed for a given arch or OS.

**Comment by mattrose** [comment]
Can someone tell me how to only apply patches to architectures that need it in the homebrew formula?  A quick check of documentation doesn't tell me that.  That way we can get this in.  I think that's the only objection?

**Comment by fxcoudert** [comment]
Wrap inside:

```
if Hardware::CPU.arm?
end
```

**Comment by yaleman** [comment]
Un-staling this in the hopes that someone will continue to work on a fix for M1 macs 😄

**Comment by mattrose** [comment]
Thanks.  I've been busy lately and this dropped off my radar, but I'll try and pick it up again v soon.

**Comment by mattrose** [comment]
Well, I guess by "very soon" I meant right now, apparently.  This was the last thing needed for merging AIUI.  Is having two different commits OK or should I squash them?

**Comment by iMichka** [comment]
Re-running a full build. You can leave 2 commits, we will auto-squash them before merging.

**Comment by mattrose** [comment]
None of the failures look to be related to my changes.

**Review by iMichka — APPROVED** [review]

**Comment by iMichka** [comment]
`dynare` failure is unrelated. That test is flaky and we should try to simplify it.

**Pull request closed without merging**